README
POLYGLOT(6)                                                        POLYGLOT(6)



NAME
       PolyGlot -  Winboard protocol to UCI protocol adapter
                -  book engine for Polyglot books
                -  a collection of utilities for creating and analyzing
       opening books
                -  a utility for analyzing epd files
                -  a perft counter

SYNOPSIS
       polyglot [configfile] [-noini] [-ec engine] [-ed enginedirectory] [-en
       enginename] [-log true/false] [-lf logfile] [-pg <name>=<value>]* [-uci
       <name>=<value>]*

       polyglot make-book [-pgn inputfile] [-bin outputfile] [-max-ply ply]
       [-min-game games] [-min-score score] [-only-white] [-only-black]
       [-uniform]

       polyglot merge-book -in1 inputfile1 -in2 inputfile2 [-out outputfile]

       polyglot info-book [-bin inputfile] [-exact]

       polyglot dump-book [-bin inputfile] -color color [-out outputfile]

       polyglot [configfile] epd-test [engineoptions] [-epd inputfile]
       [-min-depth depth] [-max-depth depth] [-max-time time] [-depth-delta
       delta]

       polyglot perft [-fen fen] [-max-depth depth]

DESCRIPTION
   PolyGlot as adapter and book engine
       PolyGlot is a "UCI adapter".  It connects a GUI interface (such as
       XBoard, Winboard, Arena or Chessbase) to a UCI chess engine.

       By specifying an opening book (in PolyGlot book format) chess engines
       can transparently use such books.

       PolyGlot understands the two main GUI protocols: UCI and xboard.
       Normally the protocol will be auto detected but this can be overridden
       in the configuration file.

       In xboard mode PolyGlot fully translates between the xboard and UCI
       protocols.  In addition it tries to solve known problems with other
       adapters.  For instance, it detects and reports draws by fifty-move
       rule, repetition, etc ... It also supports Chess960.

       When in UCI mode PolyGlot mostly passes commands from the GUI to the
       engine and vice versa, except that it will play book moves on behalf of
       the engine when the occasion arises.

       The engine options are exported as UCI options in UCI mode and as
       "feature option=" commands in xboard mode. The latter form an extension
       of the xboard protocol as defined by H.G. Muller.

       Options which normally appear in the [PolyGlot] section of the config
       file (see below) are exported as options with their name prefixed by
       "Polyglot". This makes it easy to filter them in the GUI.

       NOTE: Not all options are exported, only those that make sense in the
       given mode.

   Book making utilities
       PolyGlot supports the "PolyGlot opening book format". This is the
       defacto standard non-proprietary opening book format. It is fully
       documented here

       http://hgm.nubati.net/book_format.html

       Roughly speaking a PolyGlot opening book is a collection of triples
       (position, move, weight). A "position" is represented by a 64-bit
       Zobrist hash key. The weight is proportional to the probability the
       move should be played.

       Other opening book formats such as ChessBase's .ctg format and Arena's
       .abk format are undocumented and proprietary. They can only be used by
       their own GUIs.

       PolyGlot can compile a pgn file into a binary PolyGlot book and
       furthermore it can merge two such binary books into a third one.

       PolyGlot can also extract some useful information from PolyGlot books.
       The utility "dump-book" dumps the "lines" in a book for a given color.
       By definition a line is a sequence of moves (from the starting
       position) in which the given color makes only book moves and the other
       color makes arbitrary moves (i.e. not necessarily book moves).

       Since a PolyGlot book is built up from positions and not lines there
       may be (and there usually are) many positions in the book that are not
       on a "line" as defined in the previous paragraph. It is convenient to
       call such positions "isolated" positions. The utility "info-book"
       counts such isolated positions.

       Some of the isolated positions are provably unreachable and they could
       in principle be deleted from the book. For example if a book contains
       only the move "e4" in the starting position but also the position after
       "d4 d5" then this last position is provably unreachable since it
       requires white to make a non-book move when a book move is available.
       Such situations arise frequently from the priority rules in merging
       books.

       Unfortunately not all isolated positions are provably unreachable and
       it is difficult to identify the latter. If invoked with "-exact" the
       utility info-book will attempt to count the isolated positions which
       require a player to make a non-book move when a book move is available.
       Due to the possibility of transpositions this is not a fool proof
       method.

   Epd test mode
       In epd test mode, PolyGlot will search positions in an epd file and
       record the number of times the right best move was found.  The
       arguments specify when to stop the search in any given position.

   Perft counts
       A perft count is the number of legal move sequence in a given position
       up to a given depth. PolyGlot can perform such perft counts. It is
       however much slower than other more dedicated programs.

OPTIONS
       When PolyGlot is invoked as an adapter of in epd-test mode it gets its
       options from a config file and then from the command line.  The default
       config file is "polyglot.ini" but an alternative one may be optionally
       included as first argument. The config file format is described below.

       The following engine options may be specified on the command line.

       -noini
           Do not use a config file, even if one was specified on the command
           line.

       -pg The argument is a string of the form <name>=<value>. This option
           will set the Polyglot option <name> to <value>.

       -uci
           The argument is a string of the form <name>=<value>. This option
           will set the engine option <name> to <value>.

       -ec This is an alias for -pg "EngineCommand=<value>"

       -ed This is an alias for -pg "EngineDir=<value>"

       -en This is an alias for -pg "EngineName=<value>"

       -log (default: false)
           This is an alias for -pg "Log=<value>"

       -lf (default: "polyglot.log")
           This is an alias for -pg "LogFile=<value>".

       -wb (default: "true")
           This is an alias for -pg "OnlyWbOptions=<value>".

       When invoked as

   polyglot make-book
       PolyGlot supports the following options

       -pgn (default: "book.pgn")
           Input file in pgn format.

       -bin (default: "book.bin")
           Output file in PolyGlot format.

       -max-ply (default: 1024)
           Specifies the maximum ply-depth of lines included in the book.

       -min-game (default: 3)
           Specifies the minimum number of games that have to contain this
           move for it to be included in the book.

       -min-score (default: 0.0)
           Specifies the minimum score (or weight) this move should have
           received for it to  be included in the book. The score is
           2*(wins)+(draws), globally scaled to fit into 16 bits.

       -only-white
           Include only moves for white in the book.

       -only-black
           Include only moves for black in the book.

       -uniform
           Set all weights to 1. In other words, all moves will be selected
           with equal probability.

       When invoked as

   polyglot merge-book
       PolyGlot supports the following options

       -in1
           First input file (in PolyGlot book format).

       -in2
           Second input file (in PolyGlot book format).

       -out (default: out.bin)
           Output file (in PolyGlot book format).

       Input files are not symmetrical, "in1" has priority over "in2". In
       other words when a position occurs both in "in1" and "in2" only the
       moves and weights from "in1" will be retained in "out".

       When invoked as

   polyglot dump-book
       PolyGlot supports the following options

       -bin (default: book.bin)
           Input file in PolyGlot book format.

       -color
           The color for whom to generate the lines.

       -out (default: book_<color>.txt)
           The name of the output file.

       When invoked as

   polyglot info-book
       PolyGlot supports the following options

       -bin (default: book.bin)
           Input file in PolyGlot book format.

       -exact
           Attempt to count the provably unreachable positions among the
           isolated ones.  Note that this takes a very long time.

       When invoked as

   polyglot epd-test
       (possibly with a config file as first argument) PolyGlot supports
       besides the generic options described above the following additional
       options.

       -max-depth (default: 63)
           Unconditionally stop the search when this depth has been reached.

       -max-time (default: 5.0)
           Unconditionally stop the seach after this amount of time.

       -depth-delta (default: 3)
           Stop the search if the solution as been found and the best move has
           been constant for this many depths, on condition that the mininal
           depth and minimal time have been reached.

       -min-depth (default: 8)
           Minimal search depth when the search is stopped using
           "-depth-delta".

       -min-time (default: 1.0)
           Minimal search time when the search is stopped using
           "-depth-delta".

       When invoked as

   polyglot perft
       PolyGlot supports the following options

       -fen (default: starting position)
           Fen at which to start searching.

       -max-depth (default: 1)
           Maximum depth to search.

CONFIG FILE FORMAT
       There should be a different config file for each engine.

       The config file is in the traditional INI format.

           [PolyGLot]
           option = value
           ...
           [Engine]
           option = value
           ...

       The characters "#" and ";" serve as comment characters.

       Initial and final white space is stripped from option names and values.
       If you need to use characters which have a special meaning to PolyGlot
       (these are ';#[]=\') you should quote them by preceding them with '\'.
       "Quoting" other characters in this way has no effect. In particular the
       use of '\' as a path separator in windows should normally not affected.

   [PolyGlot] section
       This section is used by PolyGlot only.  The engine is unaware of these
       options.  The list of available options is detailed below.

       EngineName (default: UCI name)
           This is the name that will appear in the GUI.  It is cosmetic only.
           You can use different names for tweaked versions of the same
           engine.

       EngineDir (default: ".")
           Full path of the directory where the engine is installed.  You can
           use "." (without the quotes) if you know that PolyGlot will be
           launched in the engine directory or the engine is in the "path" and
           does not need any data file.

       EngineCommand
           Put here the name of the engine executable file.  You can also add
           command-line arguments.  Path searching is used and the current
           directory will be "EngineDir". On Linux the EngineCommand is passed
           to wordexp so that shell quoting rules and expansions are applied.
           On Windows the EngineCommand is simply passed to CreateProcess
           which does its own shell like processing.

       SettingsDir (default: $HOME/.polyglot on Linux; ".\_PG" on Windows)
           The directory where ini files are stored for engines that are
           started with -noini. Such ini files may be created by pushing the
           "Save" button in the Engine settings dialog in WB/XB 4.4.0 and
           higher. As a special exception (for WB/XB 4.4.0 compatibility) this
           directory is also used in case PolyGlot is started with config
           files named "polyglot_1st.ini" or "polyglot_2nd.ini".

       Log (default: false)
           Whether PolyGlot should log all transactions with the interface and
           the engine.  This should be necessary only to locate problems.

       LogFile (default: polyglot.log)
           The name of the log file.  Note that it is put where PolyGlot was
           launched from, not into the engine directory.

           WARNING: Log files are not cleared between sessions, and can become
           very large.  It is safe to remove them though.

       Resign (default: false)
           Set this to "true" if you want PolyGlot to resign on behalf of the
           engine.

           NOTE: Some engines display buggy scores from time to time although
           the best move is correct.  Use this option only if you know what
           you are doing (e.g. you always check the final position of games).

       ResignMoves (default: 3)
           Number of consecutive moves with "resign" score (see below) before
           PolyGlot resigns for the engine.  Positions with only one legal
           move are ignored.

       ResignScore (default: 600)
           This is the score in centipawns that will trigger resign
           "counting".

       QueenNeverResigns (default: true)
           Suppresses resigning irrespective of the score, as long as the
           engine still has a Queen (hoping for a perpetual).

       HandleDraws (default: false)
           When set Polyglot can accept draw offers on behalf of the engine,
           when the last reported engine score is below Polyglot's contempt.

       ContemptScore (default: 30)
           This is the score in centipawns that the engine must minimally see
           in its own disadvantage before Polyglot will accept a draw offer.

       ShowPonder (default: true)
           Show search information during engine pondering.  Turning this off
           might be better for interactive use in some interfaces.

       ShowTbHits (default: false)
           Reports extra info on selective depth, search speed (knps) and
           tablebase hits at the beginning of the PV field in the thinking
           output of WinBoard protocol, just before the real PV.

       ScoreWhite (default: true)
           Report score from white's point of view in xboard mode. (Wrecks
           XBoard...).

       KibitzMove (default: false)
           Whether to kibitz when playing a move.

       KibitzPV (default: false)
           Whether to kibitz when the PV is changed (new iteration or new best
           move).

       KibitzCommand (default: "tellall")
           xboard command to use for kibitzing, normally "tellall" for
           kibitzing or "tellothers" for whispering.

       KibitzDelay (default: 5)
           How many seconds to wait before starting kibitzing.  This has an
           effect only if "KibitzPV" is selected, move kibitzes are always
           sent regardless of the delay.

       KibitzInterval (default: 0)
           This is another form of throttling. PolyGlot will usually wait this
           many seconds before doing the next kibitz.

       UCI (default: false)
           If true PolyGlot will not understand xboard commands.

       MateScore (default: 0)
           Mate score reported to GUI when in xboard mode.  When set to 0 it
           uses the new CECP convention of reporting 100000 + N for mate in N
           moves, and -(100000 + N) for mated in N.

       Book (default: false)
           Indicates whether a PolyGlot book should be used.  This has no
           effect on the engine own book (which can be controlled with the UCI
           option "OwnBook" in the [Engine] section).  In particular, it is
           possible to use both a PolyGlot book and an engine book.  In that
           case, the engine book will be used whenever PolyGlot is out of
           book.  Remember that PolyGlot is unaware of whether the engine is
           itself using a book or not.

       BookFile (default: book.bin)
           The name of the (binary) book file.  Note that PolyGlot will look
           for it in the directory it was launched from, not in the engine
           directory.  Of course, full path can be used in which case the
           current directory does not matter.

       BookRandom (default: true)
           Select moves according to their weights in the book. If false the
           move with the highest weight is selected.

       BookLearn (default: false)
           Switches on updating of the 'learn fields' of the opening book,
           which for each move record the number of times the move was played,
           and the number of half-points scored with it.  It also switches on
           making use of this information already in the book, for altering
           the move-selection probabilities from those specified by the book
           weights.  (Rarely played moves will be favored, moves with a
           significant poor performance will be shunned.)  This option must be
           set before the book is specified, or the updating of the learn info
           will not work!

       BookDepth (default: 256)
           Stop using the book after this number of moves.

       BookTreshold (default: 5)
           Do not play moves with a weight (probability) lower than this (in
           per mil).

       UseNice (default: false)
           Run the engine at nice level 5, or "NiceValue" if it set.  On some
           operating systems it may be necessary to run the engine at lower
           priority for it to be responsive to commands from PolyGlot while
           searching.

       NiceValue (default: 5)
           Nice levels go from -20 to 20 with 20 being the lowest priority.
           On Unix only root can set negative nice levels. On Windows the
           standard Win32 priority levels are mapped in a sensible way to Unix
           nice levels.

       Affinity (default: -1)
           This a bit vector in which each bit represents the processors that

ChangeLog
=========2.0.4==================
- Fix move choice using learn info (when BookLearn is on).
- Use new XBoard mate-score convention by default.
- Support UCI_PlayByNodes option / CECP nps command, and simulate this option using 'go nodes' when the engine does not support it.
=========2.0.3==================
- Bugfix, removing a debug printf which would forward all engine output to the GUI as comment. (Which would cause problems if the line contained "not found".)
=========2.0.2==================
- The Polyglot Affinity option was activated to set the proces affinity mask on Windows, and turned into a string option to allow it to be given in hexadecimal.
- The makefile for building with MSVC was repaired, and several docs updated.
- The default name for engines that do not emit an "id name" was changed to "unknown".
- Fail-low and fail-high info provided by the engine is realyed in thinking output as trailing ? and ! on the PV.
- The UCI option SyzygyPath is recognized as special, and connected to the XBoard 'egtpath' command.
- Tablebase hits, selective depth and nodes/sec can be shown in thinking output, subject to the Polyglot option ShowTbHits.
- XBoard commands wrapped as info strings ("info string xboard XXX") will now be passed to the GUI (as "XXX").
- Hash size is now capped at the engine-specified maximum, even when the XBoard memory command allows more tha that.
- The exclude-moves feature now fully supports the WB protocol extension for this (exclude MOVE|all, include MOVE|all).
=========2.0.1==================
- This is the first of a fork initiated by HGM. It starts by reverting the previous book-learning change, as the reasons for it seem invalid. People do not use book learning for making better opening books, but for adapting a book to one specific engine, and provide diversity in on-line play, to prevent human opponents from exploiting a likely line they once won by repeating it again and again.
- Polyglot was now also made to use the learn info it updates, for promoting diversity and avoidance of lines with a poor result record.
- An option to refrain from resigning while you still have a Queen (QueenNeverResigns) was added
- Option to accept draw offers (HandleDraws and ContemptScore) were added.
- A first onset towards an exclude-moves feature was made.
========1.4.67b=================
- Disable "book learning". Polyglot saves learning information but does not use it itself, probably for a reason. I believe the booklearning concept of Polyglot is fundamentally broken. Learning information is engine specific so it should not be saved in the opening book (it might be saved elsewhere). There are other issues as well, such as copyright and the possibility of the proliferation of polluted books. Opening books can be improved with information from games by using the merge command. This is a form of off line learning which is probably much more effective since one has strict control over what kind of information is added to the book.
========1.4.66b=================
- Fix a printf 32 vs 64 bit format (crash) bug which manifested itself on windows.
========1.4.65b=================
- Bugfix: dump-book with -color black generated the file book_white.txt.
========1.4.64b=================
- Reduce movetime in the implementation of the fixed time per move command, to to account for differences in the wording of the protocols. 
========1.4.63b=================
- Give ponder move as hint.
- Support for egtpath gaviota (this is a bit hacky right now).
========1.4.62b=================
- StringSize was not everywhere the same in Polyglot. This could lead to a buffer overflow in case of very long PV's. 
========1.4.61b=================
- uci_isready replaced by uci_isready_sync. Polyglot should not send commands to the engine while it is syncing.
========1.4.60b=================
- The result string after an illegal move now shows the actual attempted move.
========1.4.59b=================
- Bugfix: the changes in 1.4.57b created a regression where the input buffer of polyglot could overflow with engines producing lots of output very fast. 
========1.4.58b=================
- Implementation of BookDepth in UCI mode.
- Small corrections to the manpage.
========1.4.57b=================
- First attempt at killing engines that do not react to "quit". 
========1.4.56b=================
- Better handling of non-existing engine command on Linux. No changes on Windows.
=========1.4.55b================
- Suppression of some error dialogs.
- Small delay in between final error message and actually quitting.
=========1.4.54b================
- More graceful handling of engine crashes.
- New EngineCommand is now passed through wordexp on Linux.
=========1.4.53b================
- Multipv code by HGM. 
=========1.4.52b================
- Reexport SyncStop. It is apparently necessary for the Aquarium Book Adapter.
=========1.4.51b================
- Support for new nps command. Code provided by Harm Geert Muller. 
=========1.4.50b================
- Change in architecture. Persistence is disabled. The user is required to push the "Save" button explicitly to save the settings. What is retained is that in case PG is invoked with -noini (or as a special exception for WB 4.4.0 when the config files are polyglot_1st.ini or polyglot_2nd.ini) then PG uses a default config file whose name is derived from the engine name. 
- New option: OnlyWbOptions (default: true). Restrict the options that PG sends to those that are potentially useful for WinBoard.
=========1.4.47b================
- Try to detect if the engine is not an UCI engine.
=========1.4.46b================
- The default persistence directory on Windows is now _PG. The names of the engine specific ini files are mangled so as not to contain spaces. 
- PG now tries to guess if a string option is really a file or a path (like in the Arena GUI). 
=========1.4.45b================

option.c

// option.c

// includes

#include <stdlib.h>
#include <string.h>

#include "option.h"
#include "util.h"


// defines

#define NNB { NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL }

// constants

static const bool UseDebug = FALSE;

// variables


option_list_t Option[1];

option_t DefaultOptions[] = {    
   // options

    { "SettingsFile",     "file","0","0",       "polyglot.ini", NULL,0,NNB,  PG|XBOARD|XBSEL}, 

    { "SettingsDir",      "path","0","0",       "<empty>"   , NULL,0,NNB,  PG},

    { "OnlyWbOptions",    "check","0","0",      "true"      , NULL,0,NNB,  PG|XBOARD},

    { "EngineName",       "string","0","0",     "<empty>"   , NULL,0,NNB,  PG}, 
    { "EngineDir",        "path","0","0",       "."         , NULL,0,NNB,  PG}, 
    { "EngineCommand",    "string","0","0",     "<empty>"   , NULL,0,NNB,  PG}, 

    { "Log",              "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD|XBSEL|UCI}, 
    { "LogFile",          "file","0","0",       "polyglot.log", NULL,0,NNB,  PG|XBOARD|XBSEL|UCI}, 

    { "UCI",              "check","0","0",      "false"     , NULL,0,NNB,  PG}, 

    { "UseNice",          "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD|UCI}, 
    { "NiceValue",        "spin", "0","20",     "5"         , NULL,0,NNB,  PG|XBOARD|UCI}, 

    { "Resign",           "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD|XBSEL}, 
    { "ResignMoves",      "spin","0","10000",    "3"        , NULL,0,NNB,  PG|XBOARD|XBSEL}, 
    { "QueenNeverResigns","check","0","0",      "true"      , NULL,0,NNB,  PG|XBOARD|XBSEL}, 
    { "ResignScore",      "spin","0","10000",   "600"       , NULL,0,NNB,  PG|XBOARD|XBSEL}, 
    { "HandleDraws",      "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD|XBSEL}, 
    { "ContemptScore",    "spin","0","10000",   "30"        , NULL,0,NNB,  PG|XBOARD|XBSEL}, 

    { "MateScore",        "spin","0","100000",  "0"         , NULL,0,NNB,  PG|XBOARD}, 

    { "Book",             "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD|XBSEL|UCI}, 
    { "BookFile",         "file","0","0",       "book.bin"  , NULL,0,NNB,  PG|XBOARD|XBSEL|UCI}, 

    { "BookRandom",       "check","0","0",      "true"      , NULL,0,NNB,  PG|XBOARD|XBSEL|UCI}, 
    { "BookDepth",        "spin","0","256",     "256"       , NULL,0,NNB,  PG|XBOARD|XBSEL|UCI}, 
    { "BookTreshold",     "spin","0","1000",    "5"         , NULL,0,NNB,  PG|XBOARD|UCI}, 
    { "BookLearn",        "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD}, 
    { "ShowTbHits",       "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD|XBSEL}, 

    { "KibitzMove",       "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD}, 
    { "KibitzPV",         "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD}, 

    { "KibitzCommand",    "string","0","0",     "tellall"   , NULL,0,NNB,  PG|XBOARD}, 
    { "KibitzDelay",      "spin","0","1000",    "5"         , NULL,0,NNB,  PG|XBOARD}, 
    { "KibitzInterval",   "spin","0","1000",    "0"         , NULL,0,NNB,  PG|XBOARD}, 

    { "ShowPonder",       "check","0","0",      "true"      , NULL,0,NNB,  PG|XBOARD}, 
    { "ScoreWhite",       "check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD}, 

    { "STFudge",          "spin","0","1000",      "20"      , NULL,0,NNB,  PG|XBOARD|XBSEL}, 

   // work-arounds

    { "UCIVersion",       "spin","1","2",       "2"         , NULL,0,NNB,  PG|XBOARD}, 
    { "CanPonder",        "check","1","2",      "false"     , NULL,0,NNB,  PG|XBOARD}, 
    { "SyncStop",         "check","1","2",      "false"     , NULL,0,NNB,  PG|XBOARD|XBSEL}, 
    { "Affinity",         "spin","0","0",       "FFFFFFFF"  , NULL,0,NNB,  PG|XBOARD}, 
    { "RepeatPV",         "check","0","0",      "true"      , NULL,0,NNB,  PG|XBOARD},
    { "PromoteWorkAround","check","0","0",      "false"     , NULL,0,NNB,  PG|XBOARD},

   // internal

    { "Chess960",         "check","0","0",      "false"     , NULL,0,NNB,  PG}, 

   // These options flag various hacks in the source to work around
   // WB quirks. They will eventually all be set to false. Probably
   // in 4.5.0
    
    { "WbWorkArounds",    "check","0","0",      "true"      , NULL,0,NNB,  PG}, 
    { "WbWorkArounds2",   "check","0","0",      "false"     , NULL,0,NNB,  PG}, 
    { "WbWorkArounds3",   "check","0","0",      "true"      , NULL,0,NNB,  PG}, 

    // Buttons

    { "Save",             "save","0","0",       "false"     , NULL,0,NNB,  PG|XBOARD|XBSEL},

    // Sentinel
    
    { NULL,               NULL,"0","0",         NULL        , NULL,0,NNB,  0},

};


// functions

// option_is_ok()

bool option_is_ok(const option_list_t *option) {
    if(option->option_nb<0 || option->option_nb>=OptionNb){
        return FALSE;
    }
    return TRUE;
}

// option_init_pg()

void option_init_pg() {

    int i;
    option_t *p=DefaultOptions;
    char *home_dir;
    char SettingsDir[StringSize];
    
    option_init(Option);
    while(p){
        if(p->name){
            option_insert(Option,p);
            p++;
        }else{
            break;
        }
    }
    for(i=0;i<Option->option_nb;i++){
        Option->options[i].value=my_strdup(Option->options[i].default_);
    }
#ifndef _WIN32
    home_dir=getenv("HOME");
    if(!home_dir){
        home_dir=".";
    }
    snprintf(SettingsDir,sizeof(SettingsDir),"%s/.polyglot",home_dir);
    SettingsDir[sizeof(SettingsDir)-1]='\0';
#else
    sprintf(SettingsDir,".\\_PG");
#endif
    option_set(Option,"SettingsDir",SettingsDir);
    option_set_default(Option,"SettingsDir",SettingsDir);
}

// option_init()

void option_init(option_list_t *option){
    ASSERT(option!=NULL);
    option->option_nb=0;
    option->iter=0;
    memset(option->options,0,sizeof(option->options));
}

// option_insert()

void option_insert(option_list_t *option, option_t *new_option){
    int i;
    option_t *opt;
    ASSERT(option!=NULL);
    ASSERT(new_option!=NULL);
    ASSERT(new_option->name!=NULL);
    opt=option_find(option,new_option->name);
    if(!opt){
        opt=&option->options[option->option_nb];
        option->option_nb++;
    }
    if(option->option_nb>=OptionNb){
        my_fatal("option_insert(): option list overflow\n");
    }
    if(new_option->name)     my_string_set(&opt->name,     new_option->name);
    if(new_option->value)    my_string_set(&opt->value,    new_option->value);
    if(new_option->min)      my_string_set(&opt->min,      new_option->min);
    if(new_option->max)      my_string_set(&opt->max,      new_option->max);
    if(new_option->default_) my_string_set(&opt->default_, new_option->default_);
    if(new_option->type)     my_string_set(&opt->type,     new_option->type);
    opt->var_nb=new_option->var_nb;
    for(i=0;i<new_option->var_nb;i++){
        my_string_set(&opt->var[i], new_option->var[i]);
    }
    opt->mode=new_option->mode;
}

// option_set()

bool option_set(option_list_t *option, 
                const char name[], 
                const char value[]) {

   option_t * opt;
   ASSERT(option!=NULL);
   ASSERT(name!=NULL);
   ASSERT(value!=NULL);

   opt = option_find(option,name);
   if (opt == NULL) return FALSE;

   if(my_string_case_equal(opt->type,"check")){
      value=(my_string_equal(value,"1")||
	     my_string_case_equal(value,"true"))?"true":"false";
   }

   my_string_set(&opt->value,value);

   if (UseDebug) my_log("POLYGLOT OPTION SET \"%s\" -> \"%s\"\n",opt->name,opt->value);

   return TRUE;
}

// option_set_default()

bool option_set_default(option_list_t *option,
                           const char name[], 
                           const char value[]) {

   option_t * opt;
   ASSERT(name!=NULL);
   ASSERT(value!=NULL);

   opt = option_find(option,name);
   if (opt == NULL) return FALSE;

   if(my_string_case_equal(opt->type,"check")){
      value=(my_string_equal(value,"1")||
	     my_string_case_equal(value,"true"))?"true":"false";
   }

   my_string_set(&opt->default_,value);

   if (UseDebug) my_log("POLYGLOT OPTION DEFAULT SET \"%s\" -> \"%s\"\n",opt->name,opt->default_);

   return TRUE;
}

// option_get()

const char * option_get(option_list_t *option, const char name[]) {

   option_t * opt;

   ASSERT(name!=NULL);

   opt = option_find(option,name);
   if (opt == NULL) my_fatal("option_get(): unknown option \"%s\"\n",name);
   if (UseDebug) my_log("POLYGLOT OPTION GET \"%s\" -> \"%s\"\n",opt->name,opt->value);

   return opt->value;
}

// option_get_default()

const char * option_get_default(option_list_t *option, const char name[]) {

   option_t * opt;

   ASSERT(name!=NULL);

   opt = option_find(option,name);
   if (opt == NULL) my_fatal("option_get(): unknown option \"%s\"\n",name);

   if (UseDebug) my_log("POLYGLOT OPTION GET \"%s\" -> \"%s\"\n",opt->name,opt->value);

   return opt->default_;
}

// option_get_bool()

bool option_get_bool(option_list_t *option, const char name[]) {

   const char * value;

   value = option_get(option,name);

   if (FALSE) {
   } else if (my_string_case_equal(value,"true") || my_string_case_equal(value,"yes") || my_string_equal(value,"1")) {
      return TRUE;
   } else if (my_string_case_equal(value,"false") || my_string_case_equal(value,"no") || my_string_equal(value,"0")) {
      return FALSE;
   }

   ASSERT(FALSE);

   return FALSE;
}


// option_get_double()

double option_get_double(option_list_t *option, const char name[]) {

   const char * value;

   value = option_get(option,name);

   return atof(value);
}

// option_get_int()

int option_get_int(option_list_t *option, const char name[]) {

   const char * value;

   value = option_get(option,name);

   return atoi(value);
}

// option_get_string()

const char * option_get_string(option_list_t *option, const char name[]) {

   const char * value;

   value = option_get(option,name);

   return value;
}

// option_find()

option_t * option_find(option_list_t *option, const char name[]) {

   option_t * opt;
   int i;

   ASSERT(name!=NULL);
   for (i=0; i<option->option_nb; i++){
       opt=option->options+i;
       if (my_string_case_equal(opt->name,name)){
           return opt;
       }
   }
   
   return NULL;
}

// option_start_iter()

void option_start_iter(option_list_t *option){
    option->iter=0;
}

// option_next()

option_t * option_next(option_list_t *option){
    ASSERT(option->iter<=option->option_nb);
    if(option->iter==option->option_nb){
        return NULL;
    }
    return &option->options[option->iter++];
        
}

void option_free(option_t *option){
      int i;
      my_string_clear(&option->name);
      my_string_clear(&option->type);
      my_string_clear(&option->min);
      my_string_clear(&option->max);
      my_string_clear(&option->default_);
      my_string_clear(&option->value);
      for(i=0;i<option->var_nb;i++){
         my_string_clear(&option->var[i]);
      }
      option->var_nb=0;
      option->mode=0;
}

// option_clear()

void option_clear(option_list_t *option){
    int i;
    for (i = 0; i < option->option_nb; i++) {
        option_free(option->options+i);
   }
   option->option_nb=0;
}

// option_from_ini()

void option_from_ini(option_list_t *option,
                     ini_t *ini,
                     const char *section){
    ini_entry_t *entry;
    ini_start_iter(ini);
    while((entry=ini_next(ini))){
        option_set(option,entry->name,entry->value);
        option_set_default(option,entry->name,entry->value);
    }
}

// end of option.cpp

